### PR TITLE
PUBDEV-8312: Speed-up GBM by optimizing histogram building

### PR DESCRIPTION
**Pull request by michalkurka**
1. Optimize binning for small-cardinality categoricals
2. Collocate response data by tree-node for faster random access in CPU cache
3. Stop model building if new trees cannot bring an improvement (won't change the model)

### CODE REVIEW

File: h2o-algos/src/main/java/hex/tree/DHistogram.java [hunk 1]
```diff
@@ -331,6 +341,8 @@ else if (_histoType== SharedTreeModel.SharedTreeParameters.HistogramType.Quantil
         _splitPtsFloat[i] = (float) binAt(i);
       }
     }
+    assert !_intOpt || _splitPts == null : "Integer-optimization cannot be enabled when split points are defined";
```
Comment (maurever) [inline h2o-algos/src/main/java/hex/tree/DHistogram.java:344]: You check  `_splitPts == null` here in the assert, but I can't see this check in `useIntOpt()` method. There should be something like (similar to constraints):

```
    if (_splitPts != null) {
      return false;
    }
```

Or am I missing something?

**Review by maurever — COMMENTED** [review]
Looks good to me. I left just one comment. Also, tests are missing. Otherwise cool improvement!

File: h2o-algos/src/main/java/hex/tree/DHistogram.java (hunk 1 above)
Comment (maurever) [reply h2o-algos/src/main/java/hex/tree/DHistogram.java:344]: I got it, the _splitPts is initialized later (before this assert).

**Review by maurever — APPROVED** [review]
Tests look good. Thank you @michalkurka.

**Comment by michalkurka** [comment]
@maurever thank you for your review and the comments you send

**Pull request merged**